json_rpc/src/filters.rs

```diff
@@ -0,0 +1,211 @@
+                    if name.as_str() == CONTENT_TYPE.as_str() {
+                        if value
+                            .as_bytes()
+                            .eq_ignore_ascii_case(CONTENT_TYPE_VALUE.as_bytes())
+                        {
+                            return Ok(());
+                        } else {
+                            trace!(content_type = ?value.to_str(), "invalid {}", CONTENT_TYPE);
+                            return Err(reject::custom(UnsupportedMediaType));
+                        }
+                    }
+                }
+                trace!("missing {}", CONTENT_TYPE);
+                Err(reject::custom(MissingContentTypeHeader))
+            }),
+        )
+        .untuple_one()
+        .and(body::content_length_limit(max_body_bytes as u64).or_else(
+            move |_rejection| async move { Err(reject::custom(BodyTooLarge(max_body_bytes))) },
+        ))
+        .boxed()
+}
+
+/// Handles parsing a JSON-RPC request from the given HTTP body, executing it using the appropriate
+/// handler, and providing a JSON-RPC response (which could be a success or failure).
+///
+/// Returns an `Err(Rejection)` only if the request is a Notification as per the JSON-RPC
+/// specification, i.e. the request doesn't contain an "id" field.  In this case, no JSON-RPC
+/// response is sent to the client.
+async fn handle_body(body: Bytes, handlers: RequestHandlers) -> Result<Response, Rejection> {
+    let response = match serde_json::from_slice::<UnvalidatedRequest>(&*body) {
+        Ok(unvalidated_request) => match Request::try_from(unvalidated_request) {
+            Ok(request) => handlers.handle_request(request).await,
+            Err(ErrorOrRejection::Error { id, error }) => {
+                debug!(?error, "got an invalid request");
+                Response::new_failure(id, error)
+            }
+            Err(ErrorOrRejection::Rejection(rejection)) => {
+                debug!(?rejection, "rejecting an invalid request");
+                return Err(rejection);
+            }
+        },
+        Err(error) => {
+            if let Ok(malformed) = serde_json::from_slice::<MalformedRequestWithId>(&*body) {
+                debug!(%error, "got an invalid request, but with 'id' field");
+                let error = Error::new(ReservedErrorCode::InvalidRequest, error.to_string());
+                Response::new_failure(malformed.id, error)
+            } else if serde_json::from_slice::<Value>(&*body).is_ok() {
+                debug!(%error, "got json, but not a request");
+                let error = Error::new(ReservedErrorCode::InvalidRequest, error.to_string());
+                Response::new_failure(Value::Null, error)
+            } else {
+                debug!(%error, "got bad json");
+                let error = Error::new(ReservedErrorCode::ParseError, error.to_string());
+                Response::new_failure(Value::Null, error)
+            }
```

Comment:
Do we need this granularity ?

Reply:
I think we do - at least that's my reading of [the error codes' meanings in the spec](https://www.jsonrpc.org/specification#error_object).